### ISSUE TITLE
onnxruntime gpu performance 5x worse than pytorch gpu performance 

### ISSUE THREAD

**Issue by snazau** [microsoft/onnxruntime#8166]
**Goal**
The end goal is to speed up the segmentation model from [segmentation_models.pytorch](https://github.com/qubvel/segmentation_models.pytorch) library using TensorRT
But since firstly I need to convert torch model into ONNX format and I faced an issue I'm here

**Describe the bug**
onnxruntime gpu performance 5x worse than pytorch gpu performance and at the same time onnxruntime cpu performance 1.5x better than pytorch cpu performance 

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.5 LTS
- ONNX Runtime installed from: pip
- ONNX Runtime version: 1.7.0
- conda version: 4.9.2
- Python version: 3.7
- PyTorch version:  1.7.0 (py3.7_cuda11.0.221_cudnn8.0.3_0)
- GPU model and memory: 1080Ti 11Gb
- PyCharm CE version: 2021.1.2

**To Reproduce**
To test I have 2 conda environments (attached in [nn-speedup.txt](https://github.com/microsoft/onnxruntime/files/6721744/nn-speedup.txt) and [nn-speedup-cpu.txt](https://github.com/microsoft/onnxruntime/files/6721745/nn-speedup-cpu.txt))

I use the following code:

    # Settings
    device = torch.device("cuda:0")
    batch_size = 16
    batches_amount = 100
    print("Device", device)
    print("Batch size", batch_size)
    print("Batch amount", batches_amount)
    print()
    
    # Input
    x = torch.randn((batch_size, 3, 256, 256))
    x = x.to(device=device)
    x_np = to_numpy(x)
    
    # Model
    import segmentation_models_pytorch as smp
    model_torch = smp.Unet('efficientnet-b5', in_channels=3, classes=2, aux_params=None)
    model_torch.encoder.set_swish(memory_efficient=False)
    model_torch = model_torch.to(device=device)
    model_torch.eval()
    
    # Save torch outputs to compare them later with ort outputs
    with torch.no_grad():
        torch_outputs = model_torch(x)
    
    # Pass runs_amount batches through the torch model
    print("Running batches with pytorch")
    torch.cuda.synchronize()
    ts = time.perf_counter()
    for i in range(batches_amount):
        with torch.no_grad():
            _ = model_torch(x)
    torch.cuda.synchronize()
    te = time.perf_counter()
    elapsed_time = te - ts
    print("{:100} {:10.7f}".format(str(batches_amount) + " model_torch(x) took", elapsed_time))
    elapsed_time = elapsed_time / batches_amount
    print("{:100} {:10.7f}".format("avg" + str(batches_amount) + " model_torch(x) took", elapsed_time))
    print()
    
    # Export torch model to ONNX format
    print("Exporting torch model to ONNx format")
    model_onnx_path = os.path.join(".", "efficientnet-b5.onnx")
    torch.onnx.export(
        model_torch,  # model being run
        x,  # model input (or a tuple for multiple inputs)
        model_onnx_path,  # where to save the model (can be a file or file-like object)
        export_params=True,  # store the trained parameter weights inside the model file
        opset_version=12,  # the ONNX version to export the model to
        do_constant_folding=True,  # whether to execute constant folding for optimization
        input_names=['input'],  # the model's input names
        output_names=['output'],  # the model's output names
        dynamic_axes={
            'input': {0: 'batch_size'},  # variable length axes
            'output': {0: 'batch_size'}
        }
    )
    
    # Clear GPU memory
    print("Clearing GPU memory")
    del model_torch
    del _
    torch.cuda.empty_cache()
    
    # Import and check ONNX model
    print("Checking model from ONNX format")
    model_onnx = onnx.load(model_onnx_path)
    onnx.checker.check_model(model_onnx)
    
    # Create InferenceSession
    print("Creating InferenceSession")
    ort_session = onnxruntime.InferenceSession(model_onnx_path)
    print("onnxruntime.get_device()", onnxruntime.get_device())
    print()
    
    # Pass runs_amount batches through the ONNX model
    print("Running batches with onnxruntime")
    ts = time.perf_counter()
    for i in range(batches_amount):
        ort_inputs = {ort_session.get_inputs()[0].name: x_np}
        _ = ort_session.run(None, ort_inputs)
    te = time.perf_counter()
    elapsed_time = te - ts
    print("{:100} {:10.7f}".format(str(batches_amount) + " ort_session.run took", elapsed_time))
    elapsed_time = elapsed_time / batches_amount
    print("{:100} {:10.7f}".format("avg" + str(batches_amount) + " ort_session.run took", elapsed_time))
    print()
    
    # Compare ONNX Runtime and PyTorch results
    ort_inputs = {ort_session.get_inputs()[0].name: x_np}
    ort_outputs = ort_session.run(None, ort_inputs)[0]
    
    print("mean abs diff =", np.mean(np.abs(to_numpy(torch_outputs) - ort_outputs)))
    np.testing.assert_allclose(to_numpy(torch_outputs), ort_outputs, rtol=1e-03, atol=1e-05)

**Results**
CPU results:

> Device cpu
> Batch size 16
> Batch amount 10
> 10 model_torch(x) took 51.9717799
> 10 ort_session.run took 19.4771159

GPU results:

> Device cuda:0
> Batch size 16
> Batch amount 100
> 100 model_torch(x) took 12.3399892
> 100 ort_session.run took 61.6554922

**Expected behavior**
I expect the ONNX model to be faster on GPU in a similar way like on CPU

**Screenshots**
Whole CPU output
![Whole CPU output](https://user-images.githubusercontent.com/30793581/123554857-6a0eb080-d782-11eb-8b21-840a0ec37c60.png)

Whole GPU output
![Whole GPU output](https://user-images.githubusercontent.com/30793581/123554862-7561dc00-d782-11eb-9a6e-e5ed4c95f8bd.png)

**Additional context**
I've also tried [that](https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html) example from PyTorch documentation and got similar results
I'm not sure about the time measurement of ort_session.run method because I think it executed asynchronously ([similarly to PyTorch](https://discuss.pytorch.org/t/copy-tensor-from-cuda-to-cpu-is-too-slow/13056/10)) but I didn't find information about correct time measurement with onnxruntime.
I will also be glad if someone shares information on setting up libraries for using TensorRT (Making work a specific combination of [CUDA, cudnn](https://www.onnxruntime.ai/docs/reference/execution-providers/CUDA-ExecutionProvider.html#requirements), [onnx and onnxruntime](https://www.onnxruntime.ai/docs/resources/compatibility.html#onnx-opset-support) was a real hell)

If you need any additional information feel free to ask

**Comment by hariharans29** [comment]
Although there might be some perf issues to be ironed out, based on the sample code you shared above, the perf comparison between Torch and ORT is not apple-to-apple.

The Torch inputs (and possibly outputs too ? - don't know where the contents of `torch_outputs` exist) seem to be copied over to the device (and outputs from the device to cpu ?) outside the latency measurement.
Input: `x = x.to(device=device)` copies it over to the device.

With ORT, the CPU input is copied over to the device and the outputs are numpy arrays (underlying memory copied over from CUDA to CPU) and these operations are included as part of the latency measurement.

Take a look at IOBinding to bind CUDA memory as inputs/outputs to ORT and this should atleast help get an apple-to-apple comparison and see the "real" delta between PyTorch and ORT - https://www.onnxruntime.ai/python/api_summary.html#id2

**Comment by snazau** [comment]
Thank you for your response @hariharans29.
You were right. The comparison wasn't correct.
I changed the code and it looks like it's all right now (hope so), but performance didn't change much.

Now inputs and outputs for PyTorch and for ORT stored on GPU
For ORT I use [scenario 3 with IOBinding](https://www.onnxruntime.ai/python/api_summary.html#id2)
Also, I'm binding inputs and outputs before for loop.

GPU results:

> Device cuda:0
> Batch size 16
> Batch amount 100
> 100 model_torch(x) took 12.3291300
> 100 ort_session.run took 61.0712226

In order not to spam with copy-paste, I attach files with code ([code.txt](https://github.com/microsoft/onnxruntime/files/6727506/code.txt)) and output ([output.txt](https://github.com/microsoft/onnxruntime/files/6727531/output.txt))

Also, I tried to use GraphOptimization with SessionOptions like it was suggested [there](https://github.com/microsoft/onnxruntime/issues/2796#issuecomment-574289547).
I end up with a bit better results (100 ort_session.run took 60.8482985), but still way slower than PyTorch.

**Comment by hariharans29** [comment]
Thanks for sharing the updated code. The IOBinding part looks correct. It is strange that the perf is exactly the same as before. I didn't expect the entire problem to go away with IOBinding but still expected that doing the device data transfer outside the measurement made the perf less worse.

We will have to investigate.

Could you please directly share the exported ONNX model (We  wouldn't have to setup Torch and do the export ourselves then) ?

Also, if you could enable verbose logging, and upload the logs that would be a bonus too.

**Comment by snazau** [comment]
It looks strange to me too. I think I have some fundamental mistake somewhere.

GPU Results with bs=1:

> Device cuda:0
> Batch size 1
> Batch amount 100
> 100 model_torch(x) took 1.8733530
> 100 ort_session.run took 7.9171801

The only thing about the logging I found was [RunOptions](https://www.onnxruntime.ai/python/api_summary.html?highlight=verbose#onnxruntime.RunOptions). Please correct me if I turned on logging in the wrong way.
I uploaded updated code, full output for bs=1 and *.onnx model to [google drive](https://drive.google.com/drive/folders/1gTFM9D0kgM0awl9moJSxJacgOz6715g9?usp=sharing)

**Comment by tianleiwu** [comment]
@snazau, I made some quick change on your script: (1) disable logging (2) add some warm up inference runs (3) change graph optimization level to extended. It seems that latency is reduced (100 run latency from 21 => 7) in my machine.  

The updated script:
[modified_code.zip](https://github.com/microsoft/onnxruntime/files/6728915/modified_code.zip)

@hariharans29, this model has Conv, could it be related to recent change in the operator?

**Comment by snazau** [comment]
@tianleiwu thank you, I tried your code with different batch sizes on my machine. It's better but still worse than PyTorch inference.

GPU Results with bs=1:

> Device cuda:0
> Batch size 1
> Batch amount 100
> 100 ort_session.run took 4.1152403

GPU Results with bs=16:

> Device cuda:0
> Batch size 16
> Batch amount 100
> 100 ort_session.run took 57.0434766

An interesting thing: I did similar (correct time measurement with IOBinding scenario 3, warm up, extended optimization level) to [an example from PyTorch Docs](https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html) (where PyTorch was faster for me too) and get nearly the same results on GPU for Pytorch and ORT (ORT ~5% faster).
So, maybe this is due to the "unfortunate" combination of hardware and network architecture?

**Comment by hariharans29** [comment]
It could be that certain Torch CUDA op implementations are faster than ORT's CUDA implementations for this architecture. We will have to profile it. I haven't looked at the model yet, but does the model have a lot of Pad nodes by any chance ? I think our Pad implementation could potentially need some perf enhancements.

**Comment by snazau** [comment]
Ok. I turned `verbose=True` during `torch.onnx.export` and according to its output I think that my model (efficientnet-b5 from [smp](https://github.com/qubvel/segmentation_models.pytorch)) has ~240 pad nodes
![image](https://user-images.githubusercontent.com/30793581/124941173-e6b74f80-e00a-11eb-8d23-7fa408c9b56e.png)

**Comment by hariharans29** [comment]
Will take a look at this next week

**Comment by hariharans29** [comment]
Hi @snazau - Please try with the referenced PR. It (together with proper warm-up, IOBinding, and highest graph optimization) should close the gap with Torch. Please re-open if you have more questions.

**Comment by snazau** [comment]
Ok, thank you!
